### DISCUSSION TITLE
Using rust structs for state management

### DISCUSSION THREAD

**Discussion by mitioshi** [slint-ui/slint#5114] [Q&A]
Hi there! Is there a way to call rust methods from slint components? 
Similarly to how in QML you can use [qmlRegisterSingletonInstance](https://doc.qt.io/qt-6/qqmlengine.html#qmlRegisterSingletonInstance) to inject an instance of a c++ class and then call its methods from qml.

I want to achieve the same in slint. How do you implement this in slint?

**Comment by bjorn** [comment]
See https://releases.slint.dev/1.5.1/docs/slint/src/recipes/recipes#global-callbacks

**Reply by mitioshi** [reply]
I would like to scope my view models closer to components. With global callback, I'd have to set every single field and method in some global scope. 
What I want to do is how qt models work here https://doc.qt.io/qt-6/qqmlengine.html#qmlRegisterSingletonInstance. 

So my questions remains the same: Is it possible to do in slint?

**Reply by bjorn** [reply]
In the docs there is a single `export global Logic`, but you can export as many globals as you want and they can also be in different files. Does that help with scoping your view models?

In any case Slint works the other way around as Qt. You define the interface in Slint files, so that the preview will work also without your C++/Rust code being available and so that a bunch of needed code can be generated. In contrast, with Qt your QML would fail to load when it can't find the imported singleton and you need to write a bunch of boilerplate code (made only somewhat bearable by macros and moc).
